test/SparqlAntlrParserTestHelpers.h

```diff
@@ -138,149 +158,212 @@ constexpr const ad_utility::Last<Current, Others...>* unwrapVariant(
+ * text has not been fully consumed by the parser. rest is expected to be the
+ * unconsumed input of the parser.
+ *
+ * @param resultOfParseAndText Parsing result
+ * @param matcher Matcher that must be fulfilled
+ * @param rest Input that is not consumed
+ */
+void expectIncompleteParse(
+    const auto& resultOfParseAndText, const string& rest, auto&& matcher,
+    ad_utility::source_location l = ad_utility::source_location::current()) {
+  auto trace = generateLocationTrace(l);
   EXPECT_THAT(resultOfParseAndText.resultOfParse_, matcher);
-  EXPECT_THAT(resultOfParseAndText.remainingText_, testing::SizeIs(length));
-}
-void expectIncompleteParse(const auto& resultOfParseAndText, auto&& matcher) {
-  expectIncompleteParse(resultOfParseAndText, 0, matcher);
+  EXPECT_THAT(resultOfParseAndText.remainingText_, testing::Eq(rest));
 }
+// _____________________________________________________________________________
+auto NumericLiteralFP =
+    [](double value) -> testing::Matcher<std::variant<int64_t, double>> {
+  return testing::VariantWith<double>(testing::DoubleEq(value));
+};
+
+auto NumericLiteralWhole =
+    [](int64_t value) -> testing::Matcher<std::variant<int64_t, double>> {
+  return testing::VariantWith<int64_t>(testing::Eq(value));
+};
+// _____________________________________________________________________________
+namespace variant_matcher {
+// Implements a matcher that checks the value of arbitrarily deeply nested
+// variants that contain a value with the last Type of the Ts. The variant may
+// also be an suffix of the Ts so to speak. E.g.
+// `MultiVariantMatcher<VarOrTerm, GraphTerm, Iri>` can match on a `Iri`,
+// `GraphTerm=std::variant<Literal, BlankNode, Iri>` and
+// `VarOrTerm=std::variant<Variable, GraphTerm>`.
+template <typename... Ts>
+class MultiVariantMatcher {
+ public:
+  explicit MultiVariantMatcher(
+      ::testing::Matcher<const ad_utility::Last<Ts...>&> matcher)
+      : matcher_(std::move(matcher)) {}
+
+  template <typename Variant>
+  bool MatchAndExplain(const Variant& value,
+                       ::testing::MatchResultListener* listener) const {
+    auto elem = unwrapVariant<Ts...>(value);
+    if (!listener->IsInterested()) {
+      return elem && matcher_.Matches(*elem);
+    }
+
+    if (!elem) {
+      *listener << "whose value is not of type '" << GetTypeName() << "'";
+      return false;
+    }
 
-// Cannot be broken down into a combination of existing googletest matchers
-// because unwrapVariant is not available as a matcher.
-MATCHER_P(IsIri, value, "") {
-  if (const auto iri = unwrapVariant<VarOrTerm, GraphTerm, Iri>(arg)) {
-    return iri->iri() == value;
+    testing::StringMatchResultListener elem_listener;
+    const bool match = matcher_.MatchAndExplain(*elem, &elem_listener);
+    *listener << "whose value " << testing::PrintToString(*elem)
+              << (match ? " matches" : " doesn't match");
+    testing::internal::PrintIfNotEmpty(elem_listener.str(), listener->stream());
```

Comment:
Except from some shortening of the code I think that the only difference should be the order. So now the message from the `MultiVariantMatcher` appears before the message of the Sub Matcher. When passing the `listener` to `MatchAndExplain` this would be probably be reversed. The method should definitely be rewritten without `internal` gtest functions.